include/irr/builtin/material_compiler/glsl/common.glsl

```diff
@@ -0,0 +1,447 @@
+{
+	return (instr.x&(1u<<INSTR_SIGMA_A_TEX_SHIFT)) != 0u;
+}
+bool instr_getWeightTexPresence(in instr_t instr)
+{
+	return (instr.x&(1u<<INSTR_WEIGHT_TEX_SHIFT)) != 0u;
+}
+bool instr_getTwosided(in instr_t instr)
+{
+	return (instr.x&(1u<<INSTR_TWOSIDED_SHIFT)) != 0u;
+}
+bool instr_getMaskFlag(in instr_t instr)
+{
+	return (instr.x&(1u<<INSTR_MASKFLAG_SHIFT)) != 0u;
+}
+bool instr_getOpacityTexPresence(in instr_t instr)
+{
+	return (instr.x&(1u<<INSTR_OPACITY_TEX_SHIFT)) != 0u;
+}
+bool instr_getTransmittanceTexPresence(in instr_t instr)
+{
+	return (instr.x&(1u<<INSTR_TRANS_TEX_SHIFT)) != 0u;
+}
+bool instr_get1stParamTexPresence(in instr_t instr)
+{
+	return (instr.x&(1u<<INSTR_1ST_PARAM_TEX_SHIFT)) != 0u;
+}
+
+//returns: x=dst, y=src1, z=src2
+uvec3 instr_decodeRegisters(in instr_t instr)
+{
+	uvec3 regs = uvec3(instr.y, (instr.y>>8), (instr.y>>16));
+	return regs & uvec3(INSTR_REG_MASK);
+}
+#define REG_DST(r)	r.x
+#define REG_SRC1(r)	r.y
+#define REG_SRC2(r)	r.z
+
+bsdf_data_t fetchBSDFDataForInstr(in instr_t instr)
+{
+	uint ix = instr_getBSDFbufOffset(instr);
+	return bsdf_buf.data[ix];
+}
+
+//remember to keep it compliant with c++ enum!!
+#define OP_DIFFUSE			0u
+#define OP_CONDUCTOR		1u
+#define OP_PLASTIC			2u
+#define OP_COATING			3u
+#define OP_MAX_BRDF			OP_COATING
+#define OP_DIFFTRANS		4u
+#define OP_DIELECTRIC		5u
+#define OP_MAX_BSDF			OP_DIELECTRIC
+#define OP_BLEND			6u
+#define OP_BUMPMAP			7u
+#define OP_SET_GEOM_NORMAL	8u
+#define OP_INVALID			9u
+#define OutColor			10u
+
+#define NDF_BECKMANN	0u
+#define NDF_GGX			1u
+#define NDF_PHONG		2u
+#define NDF_AS			3u
+
+#include <irr/builtin/glsl/bxdf/common.glsl>
+#include <irr/builtin/glsl/bxdf/brdf/specular/fresnel/fresnel.glsl>
+#include <irr/builtin/glsl/bxdf/brdf/diffuse/fresnel_correction.glsl>
+#include <irr/builtin/glsl/bxdf/brdf/diffuse/lambert.glsl>
+#include <irr/builtin/glsl/bxdf/brdf/diffuse/oren_nayar.glsl>
+#include <irr/builtin/glsl/bxdf/brdf/specular/ndf/ggx.glsl>
+#include <irr/builtin/glsl/bxdf/brdf/specular/ashikhmin_shirley.glsl>
+#include <irr/builtin/glsl/bxdf/brdf/specular/beckmann_smith.glsl>
+#include <irr/builtin/glsl/bxdf/brdf/specular/ggx.glsl>
+#include <irr/builtin/glsl/bxdf/brdf/specular/blinn_phong.glsl>
+#include <irr/builtin/glsl/bump_mapping/utils.glsl>
+
+irr_glsl_BSDFAnisotropicParams currBSDFParams;
+irr_glsl_AnisotropicViewSurfaceInteraction currInteraction;
+reg_t registers[REG_COUNT];
+reg_t prefetch_registers[TEX_PREFETCH_REG_COUNT];
+reg_t normal_registers[NORMAL_REG_COUNT];
+
+vec3 textureOrRGBconst(in uvec3 data, in bool texPresenceFlag)
+{
+	vec3 retval;
+	if (texPresenceFlag)
+		retval = prefetch_registers[data.z];
+	else
+		retval = uintBitsToFloat(data);
+	return retval;
+}
+float textureOrRconst(in uvec3 data, in bool texPresenceFlag)
+{
+	float retval;
+	if (texPresenceFlag)
+		retval = prefetch_registers[data.z].x;
+	else
+		retval = uintBitsToFloat(data.x);
+	return retval;
```

Comment:
you want to conditionally optimize these `if` statements when you know that the BSDF forest nodes will always return the same values

Reply:
also turn them into the ternary `?` operator which supposedly does not branch

Reply:
done?

Reply:
TODO